serialization_v2/serialization.py

```diff
@@ -0,0 +1,1182 @@
+            op_str = {"add": "+", "mul": "*", "truediv_right": "//", "floordiv_right": "//"}[dim.derived_from_op.kind]
+            s = f" {op_str} ".join(x.py_inline() for x in inputs_s)
+            return PyEvalCode(s, need_brackets_when_inlined=True)
+
+        # generic fallback
+        dim_type_str = self._serialize_value(type(dim), prefix="Dim", recursive=True)
+        assert isinstance(dim_type_str, PyEvalCode)
+        kwargs = {"name": repr(dim.name)}
+        if dim.kind is not None:
+            kind_s = {Dim.Types.Batch: "Batch", Dim.Types.Spatial: "Spatial", Dim.Types.Feature: "Feature"}[dim.kind]
+            kwargs["kind"] = f"{dim_type_str.py_inline()}.Types.{kind_s}"
+        return PyEvalCode(
+            f"{dim_type_str.py_inline()}"
+            f"({dim.dimension}, {', '.join(f'{key}={value}' for key, value in kwargs.items())})"
+        )
+
+    def _serialize_cached_file(self, cached_file: CachedFile, prefix: str) -> PyEvalCode:
+        # we serialize a CachedFile object, so we know that RETURNN is available for import
+        from returnn.util.file_cache import CachedFile
+
+        assert isinstance(cached_file, CachedFile)
+        cf_type_str = self._serialize_value(type(cached_file), prefix="CachedFile", recursive=True)
+        assert isinstance(cf_type_str, PyEvalCode)
+        assert isinstance(cached_file.filename, (str, Path))
+        path = self._serialize_value(cached_file.filename, prefix=f"{prefix}_filename", recursive=True)
+        return PyEvalCode(f"{cf_type_str.py_inline()}({path.py_inline()})")
+
+    def _serialize_via_repr(self, value: Any, need_brackets_when_inlined: bool) -> PyEvalCode:
+        assert isinstance(value, (Checkpoint, CodeWrapper))
+        return PyEvalCode(repr(value), need_brackets_when_inlined=need_brackets_when_inlined)
+
+
+class _SerializationDependsOnNotYetSerializedOtherVarException(Exception):
+    def __init__(self, queue_item: _AssignQueueItem):
+        super().__init__(
+            f"serialization depends on not yet serialized other var:"
+            f" ({queue_item.suggested_var_name}) = {queue_item.value!r} (type {type(queue_item.value).__name__})"
+        )
+        self.queue_item = queue_item
+
+
+class SerializationError(Exception):
+    """
+    Cannot serialize this object.
+    """
+
+
+@dataclass
+class _AssignQueueItem:
+    value: Any
+    required_var_name: Optional[str] = None
+    suggested_var_name: Optional[str] = None
+
+    via_queue_item: Optional[_AssignQueueItem] = None  # for debugging
+
+    def debug_trace(self) -> str:
+        if self.via_queue_item:
+            return (
+                f"{self.required_var_name or self.suggested_var_name} <{type(self.value).__name__}>"
+                f" -> {self.via_queue_item.debug_trace()}"
+            )
+        return f"config {self.required_var_name or self.suggested_var_name} = {self.value!r}"
+
+
+@dataclass
+class PyCode(SerializerObject):
```

Comment:
Maybe this should be private?

```suggestion
class _PyCode(SerializerObject):
```